apps/s_time.c

```diff
@@ -217,10 +218,10 @@ int s_time_main(int argc, char **argv)
     /* Loop and time how long it takes to make connections */
 
     bytes_read = 0;
-    finishtime = (long)((unsigned long)time(NULL) + maxtime);
+    starttime = time(NULL);
     tm_Time_F(START);
     for (;;) {
-        if ((long)((unsigned long)time(NULL) - finishtime) > 0)
+        if (time(NULL) - starttime > maxtime)
```

Comment:
This will be undefined behaviour is `time(2)` rolls negative.  Signed integers overflowing negative isn't defined by the C standard.

Reply:
No, time(2) may not return negative number, except:
On error, ((time_t) -1) is returned, and errno is set appropriately.

Reply:
The man page differs:
```
       On  Linux,  a  call  to time() with tloc specified as NULL cannot fail
       with the error EOVERFLOW, even on ABIs where time_t is a signed 32-bit
       integer  and  the clock ticks past the time 2**31 (2038-01-19 03:14:08
       UTC, ignoring leap seconds).  (POSIX.1 permits, but does not  require,
       the EOVERFLOW error in the case where the seconds since the Epoch will
       not fit in time_t.)  Instead, the behavior on Linux is undefined  when
       the  system time is out of the time_t range.  Applications intended to
       run after 2038 should use ABIs with time_t wider than 32 bits.
```
although it's not just a negative value, it's simply undefined in a way that can't be detected.

Reply:
Yeah, that's how I understood that too, although your man page is newer than mine.
The time_t will eventually be changed to 64 bit, but long will still be possibly a 32 bit.
So we must either use time_t, where we may expect time(NULL) to return a positive number.
or be very careful when casting time_t to long.

Reply:
So this PR is not trying to fix a targets using 32-bit time_t after 2038, but instead fix issues
when 64-bit-time_t is cast to 32-bit-long values.

Reply:
I would say that if the goal is to avoid issues when 64-bit time_t is cast to 32-bit long values, "use time_t everywhere" is clearly superior to adding new casts.  The places that are changing here are not ones where the time information is exposed in a public API.